### PR TITLE
layout/barneshut: new package for calculating Barnes-Hut force approximations

### PR DESCRIPTION
**Pull request by kortschak**
WIP

Build failures are due to issues with coveralls.io lemurheavy/coveralls-public#1264.

### CODE REVIEW

**Comment by kortschak** [comment]
The figure for the aborjs planar example taking into accound the bounds changes is here.

example-space-orig

**Comment by kortschak** [comment]
Please take a look.

I think the overall package and testing are OK, but I'm not sure of whether to provide the compact construction as a option to the `New` functions. This would require digging the old `insert` and `forcesOnMassAt` methods as the compact version methods, and adding a `compact bool` field to `Plane` and `Volume` to choose which forces method to call.

File: layout/barneshut/barneshut2.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+	dir := t.bounds.quadrant(p)
+	if t.nodes[dir] == nil {
+		t.nodes[dir] = &tile{bounds: t.bounds.split(dir)}
+	}
+	t.nodes[dir].insert(p)
+}
+
+// summarize updates node masses and centers of mass.
+func (t *tile) summarize() (center Point2, mass float64) {
+	for _, d := range &t.nodes {
```
Comment (sbinet) [inline layout/barneshut/barneshut2.go]: just a quick drive-by comment:
why `&t.nodes` ? (and not just `t.nodes`)

(ditto for `bucket.summarize()` )

Comment (kortschak) [reply layout/barneshut/barneshut2.go]: The `nodes` is an array, so this avoids a copy.

Comment (sbinet) [reply layout/barneshut/barneshut2.go]: Is range really copying the array?
I knew that when one did:
```go
for i,v := range vs
```
The `v` value was being copied.
But vs? Didn't know... Are you sure?

Comment (kortschak) [reply layout/barneshut/barneshut2.go]: https://play.golang.org/p/SBTchSt2rU5

Comment (sbinet) [reply layout/barneshut/barneshut2.go]: thanks. that's subtle.

**Comment by kortschak** [comment]
*ping*

File: layout/barneshut/barneshut2.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+// Force2 is a force modeling function for interactions between p1 and p2,
+// m1 is the mass of p1 and m2 of p2. The vector v is the vector from p1 to
+// p2. The returned value is the force vector acting on p1.
+//
+// In models where the identity of particles must be known, p1 and p2 may be
+// compared. Force2 may be passed nil for p2 when the Barnes-Hut approximation
+// is being used. A nil p2 indicates that the second mass center is an
+// aggregate.
+type Force2 func(p1, p2 Particle2, m1, m2 float64, v Point2) Point2
```
[the remark this answers is not in the document]
Comment (kortschak) [reply layout/barneshut/barneshut2.go]: Yes. In the optimising case, the second particle is not real and so is nil.

[the remark this answers is not in the document]
Comment (kortschak) [reply layout/barneshut/barneshut2.go]: Looking now, I suspect that I have that wrong, since there is no way that `p2` is ever nil AFAICS..

**Comment by kortschak** [comment]
*ping*

File: layout/barneshut/barneshut2_test.go [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+
+func TestPlane(t *testing.T) {
+	const tol = 1e-15
+
+	for _, test := range planeTests {
+		var particles []Particle2
+		if test.particles != nil {
+			particles = make([]Particle2, len(test.particles))
+		}
+		for i, p := range test.particles {
```
Comment (sbinet) [inline layout/barneshut/barneshut2_test.go]: couldn't we use `copy(particles, test.particles)` here?

File: layout/barneshut/barneshut3_test.go [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+
+func TestVolume(t *testing.T) {
+	const tol = 1e-15
+
+	for _, test := range volumeTests {
+		var particles []Particle3
+		if test.particles != nil {
+			particles = make([]Particle3, len(test.particles))
+		}
+		for i, p := range test.particles {
```
Comment (sbinet) [inline layout/barneshut/barneshut3_test.go]: ditto.

File: layout/barneshut/galaxy_example_test.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+		for j, s := range stars {
+			vectors[j] = plane.ForceOn(s, theta, barneshut.Gravity2).Scale(G)
+		}
+
+		// Update positions.
+		for j, s := range stars {
+			s.move(vectors[j])
+		}
+
+		// Rendering stars is left as an exercise for
```
Comment (sbinet) [inline layout/barneshut/galaxy_example_test.go]: pun aside, that would be a great PR plot :)
(or, even, a nice animated GIF)

File: layout/barneshut/barneshut2_test.go (hunk 3 above)
Comment (kortschak) [reply layout/barneshut/barneshut2_test.go]: No, they're not the same type: `[]Particle2` and `[]particle2`. One has an interface element and the other a concrete element.

File: layout/barneshut/barneshut3_test.go (hunk 4 above)
Comment (kortschak) [reply layout/barneshut/barneshut3_test.go]: Same

**Comment by kortschak** [comment]
Do you have any views on the slightly slower by considerably less space-costly packing?

When I merge this, I'm going to leave the full history so that we can recover that if needed. In the long run I want to merge this into gonum/gonum for two things: graph layouting (hence the name of the branch) and BH t-SNE. I don't see a good location for it in that tree though. Any suggestions?

**Comment by sbinet** [comment]
> Do you have any views on the slightly slower b[ut] considerably less space-costly packing?

I am fine with that.

> I don't see a good location for it in that tree though. Any suggestions?

perhaps `gonum/graph/layout` ?

**Comment by kortschak** [comment]
I will initially use the canonical implementation, but might add back the space-compact algorithm later (with an option to use it rather than by default).

I had initially thought that it would go in graph, but BH-tSNE doesn't fit there. I'm thinking a layout top-level.

**Comment by sbinet** [comment]
> I had initially thought that it would go in graph, but BH-tSNE doesn't fit there. I'm thinking a "gonum/layout" or "gonum/spatial" top-level.

SGTM.
`scipy` has `spatial`:
```python
>>> import scipy.spatial as spa
>>> dir(spa)
['ConvexHull',
 'Delaunay',
 'HalfspaceIntersection',
 'KDTree',
 'Rectangle',
 'SphericalVoronoi',
 'Voronoi',
 'cKDTree',
 'ckdtree',
 'convex_hull_plot_2d',
 'delaunay_plot_2d',
 'distance',
 'distance_matrix',
 'division',
 'kdtree',
 'minkowski_distance',
 'minkowski_distance_p',
 'procrustes',
 'qhull',
 'transform',
 'tsearch',
 'voronoi_plot_2d']
```

**Comment by kortschak** [comment]
Excellent.

I'd like at least one other reviewer on the implementation before I merge this here. Then I can propose it there.

@btracey @vladimir-ch Do either of you have time?

**Comment by vladimir-ch** [comment]
Not earlier than Tuesday next week.

**Comment by kortschak** [comment]
This can wait.

File: layout/barneshut/barneshut2.go [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+)
+
+const (
+	ne = iota
+	se
+	sw
+	nw
+)
+
+// Point2 is a 2D point.
```
Comment (vladimir-ch) [inline layout/barneshut/barneshut2.go]: This type is used mostly for force calculations, so I think it would be clearer to call it `Vec2` or `Vector2`. Even where it is used for particle position it can be considered as a position _vector_.

Also, I think it would be nice to have `r2` and `r3` packages as in the `geo/r2,r3` packages. Then the users of this package would implement their particles against `r2.Vec` or `r3.Vec` types instead of (longer and less natural) `barneshut.Point2,3`. What do you think?

File: layout/barneshut/galaxy_example_test.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+		p[i] = s
+	}
+	vectors := make([]barneshut.Point2, len(stars))
+
+	// Make a plane to calculate approximate forces
+	plane := barneshut.Plane{Particles: p}
+
+	// Run a simulation for 100 updates.
+	for i := 0; i < 1000; i++ {
+		// Build the data structure, For small system
```
Comment (vladimir-ch) [inline layout/barneshut/galaxy_example_test.go]: Dot instead of comma, and plural in `systems` (or an `a`).

File: layout/barneshut/barneshut2.go [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+		X: (b.Min.X + b.Max.X) / 2,
+		Y: (b.Min.Y + b.Max.Y) / 2,
+	}
+	c := p.Coord2()
+	if checkBounds && (c.X < b.Min.X || b.Max.X < c.X || c.Y < b.Min.Y || b.Max.Y < c.Y) {
+		panic(fmt.Sprintf("p out of range %+v: %#v", b, p))
+	}
+	if c.X < center.X {
+		if c.Y < center.Y {
+			return nw
```
[the remark this answers is not in the document]
Comment (kortschak) [reply layout/barneshut/barneshut2.go]: It is clock oriented sectors, counting clockwise. The coordinates are going to be used in graphics (primarily), so the coords are graphics display orientation.

File: layout/barneshut/barneshut2.go [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+	p.Y *= f
+	return p
+}
+
+// Box2 is a 2D bounding box.
+type Box2 struct {
+	Min, Max Point2
+}
+
+// quadrant returns which quadrant of b that p should be placed in.
```
[the remark this answers is not in the document]
Comment (kortschak) [reply layout/barneshut/barneshut2.go]: It is. It could be rewritten to avoid the final preposition, but it's not necessary.

File: layout/barneshut/galaxy_example_test.go (hunk 7 above)
Comment (kortschak) [reply layout/barneshut/galaxy_example_test.go]: Fixed.

File: layout/barneshut/barneshut2.go (hunk 6 above)
Comment (kortschak) [reply layout/barneshut/barneshut2.go]: The vector change sounds reasonable.

For the `r2`/`r3` changes where would these be placed? @sbinet suggested that this end up in gonum/spatial, so gonum/spatial/{r2,r3}? I would do that when I bring this over there rather than in this PR.

File: layout/barneshut/barneshut2.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (vladimir-ch) [reply layout/barneshut/barneshut2.go]: Thanks.

File: layout/barneshut/barneshut2.go (hunk 6 above)
Comment (vladimir-ch) [reply layout/barneshut/barneshut2.go]: All SGTM.

File: layout/barneshut/barneshut2.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (vladimir-ch) [reply layout/barneshut/barneshut2.go]: That's what I suspected.

**Comment by kortschak** [comment]
Do you have any views about the alternative non-leaf tree packing? It's slower because it performs more force calculations, but it packs more efficiently because more internal node slots are used. It's marginally more acurate because, again more calculations are performed. I'll merge it as is but we can revisit this later if it's seen as worth following.
